src/Nodejs.cc

```diff
@@ -200,6 +200,11 @@ Instance::Instance()
     : node_isolate_data_(nullptr, node::FreeIsolateData),
       node_environment_(nullptr, node::FreeEnvironment) {}
 
+Instance::~Instance() {
+  v8::Isolate::Scope isolate_scope(GetIsolate());
+  node_environment_.reset();
```

Comment:
This is pretty obscure action-at-distance (and the `reset` is not strictly needed). Have you thought about changing the deleter argument for at least `node_environment_` (not sure whether `node_isolate_data_` can have similar issues)? With that they could look like

```cpp
  std::unique_ptr<node::IsolateData, void (*)(node::IsolateData*)> node_isolate_data_;
  std::unique_ptr<node::Environment, void (*)(node::Environment*)> node_environment_;
```

The to create them you could pass in a lambda which makes sure to set up an `isolate::Scope` when the `Free*` functions are called (currently they are initialized in two places, might make sense to pull that into a single place).

Reply:
Thanks, I like the lambda idea for `node_environment_`.